x/incentive/keeper/keeper.go

```diff
@@ -39,10 +45,267 @@ func NewKeeper(
+	foundIncentive := false
+
+	// Fetch incentive params
+	params := k.GetParams(ctx)
+
+	// Find approporiate Incentive Info
+	// Consider epochIdentifier and start time
+	// Consider epochNumber as well
+	for _, ii := range params.StakeIncentives {
+		if ii.EpochIdentifier == epochIdentifier && timestamp >= ii.StartTime.Unix() {
+			stakeIncentive = ii
+			foundIncentive = true
+			break
+		}
+	}
+
+	// Find approporiate Incentive Info
+	for _, ii := range params.LpIncentives {
+		if ii.EpochIdentifier == epochIdentifier && timestamp >= ii.StartTime.Unix() {
+			lpIncentive = ii
+		}
+	}
+
+	// return found, stake, lp incentive params
+	return foundIncentive, stakeIncentive, lpIncentive
+}
+
+// Update uncommitted token amount
+// Called back through epoch hook
+func (k Keeper) UpdateUncommittedTokens(ctx sdk.Context, epochIdentifier string) {
+	// Recalculate total committed info
+	k.UpdateTotalCommitmentInfo(ctx, epochIdentifier)
+
+	// Iterate all delegations for the specified delegator
+	// Process to increase uncomitted token amount of Eden & Eden boost
+	k.cmk.IterateCommitments(
+		ctx, epochIdentifier,
+		func(commitments ctypes.Commitments) bool {
+			// Find out active incentive params
+			foundIncentive, stakeIncentive, lpIncentive := k.FindProperIncentiveParm(ctx, epochIdentifier)
+
+			// If we don't have incentive params ready
+			if !foundIncentive {
+				return true
+			}
+
+			// Commitment owner
+			creator := commitments.Creator
+
+			// Calculate delegated amount per delegator
+			delegatedAmt := k.CalculateDelegatedAmount(ctx, creator)
+
+			// Calculate new uncommitted Eden tokens for LP, staker, and Eden token holders
+			new_uncommitted_eden_tokens := k.CalculateNewUncommittedEdenTokens(ctx, delegatedAmt, commitments, stakeIncentive, lpIncentive)
+
+			// Calculate new uncommitted Eden-Boost tokens for staker and Eden token holders
+			new_uncommitted_eden_boost_tokens := k.CalculateNewUncommittedEdenBoostTokens(ctx, delegatedAmt, commitments, epochIdentifier, stakeIncentive, lpIncentive)
+
+			// Update Commitments with new uncommitted token amounts
+			k.UpdateCommitments(ctx, creator, new_uncommitted_eden_tokens, new_uncommitted_eden_boost_tokens)
+
+			return false
+		},
+	)
+}
+
+// Calculate new Eden token amounts based on the given conditions and user's current uncommitted token balance
+func (k Keeper) CalculateNewUncommittedEdenTokens(ctx sdk.Context, delegatedAmt sdk.Int, commitments ctypes.Commitments, stakeIncentive types.IncentiveInfo, lpIncentive types.IncentiveInfo) sdk.Int {
+	// Get LP commitments - Skip for now
+	edenCommittedByLP := sdk.ZeroInt()
+
+	// Get eden commitments and eden boost commitments
+	edenCommitted := sdk.ZeroInt()
+	edenBoostCommitted := sdk.ZeroInt()
+	for _, c := range commitments.CommittedTokens {
+		// Eden committed
+		if c.Denom == types.Eden {
+			edenCommitted = c.Amount
+		}
+
+		// Eden boost committed
+		if c.Denom == types.EdenB {
+			edenBoostCommitted = c.Amount
+		}
+	}
```

Comment:
@kenta-elys there are functions for this in the commitment type:

```
func (c *Commitments) GetUncommittedAmountForDenom(denom string) sdk.Int {
	for _, token := range c.UncommittedTokens {
		if token.Denom == denom {
			return token.Amount
		}
	}
	return sdk.NewInt(0)
}

func (c *Commitments) GetCommittedAmountForDenom(denom string) sdk.Int {
	for _, token := range c.CommittedTokens {
		if token.Denom == denom {
			return token.Amount
		}
	}
	return sdk.NewInt(0)
}
```

Reply:
Got it.